lib/elixir/lib/inspect/algebra.ex

```diff
@@ -379,7 +388,30 @@ defmodule Inspect.Algebra do
   end
 
   def to_doc(arg, %Inspect.Opts{inspect_fun: fun} = opts) do
-    fun.(arg, opts)
+    try do
+      fun.(arg, opts)
+    rescue
+      caught_exception ->
+        res =
+          Inspect.inspect(arg, %{
+            opts
+            | syntax_colors: [],
+              inspect_fun: Inspect.Opts.default_inspect_fun()
+          })
+
+        res = IO.iodata_to_binary(format(res, :infinity))
+
+        message =
+          "got #{inspect(caught_exception.__struct__)} with message " <>
+            "#{inspect(Exception.message(caught_exception))} while inspecting #{res}"
+
+        exception = Inspect.Error.exception(message: message)
+
+        Inspect.inspect(exception, %{
```

Comment:
After further exploration I think this line should become:

```elixir
reraise exception, __STACKTRACE__
```

...and that we should also add another clause to the top of this rescue expression:

```elixir
e in Inspect.Error ->
  reraise e, __STACKTRACE__
```

More details available here: https://github.com/elixir-lang/elixir/pull/11248#issuecomment-919864040